### PR TITLE
Improve benchmarks

### PR DESCRIPTION
**Pull request by Kubuxu**
License: MIT

### CODE REVIEW

**Comment by Kubuxu** [comment]
Produces actually useful output over a ranges we would expect to exist in libp2p:

```
BenchmarkEvents/subs-1/emits-1/stateful-false-4         	 5000000	       390 ns/op
BenchmarkEvents/subs-1/emits-4/stateful-false-4         	 3000000	       448 ns/op
BenchmarkEvents/subs-1/emits-16/stateful-false-4        	 3000000	       475 ns/op
BenchmarkEvents/subs-1/emits-64/stateful-false-4        	 3000000	       457 ns/op
BenchmarkEvents/subs-1/emits-256/stateful-false-4       	 3000000	       459 ns/op
BenchmarkEvents/subs-4/emits-1/stateful-false-4         	 1000000	      1449 ns/op
BenchmarkEvents/subs-4/emits-4/stateful-false-4         	 1000000	      1436 ns/op
BenchmarkEvents/subs-4/emits-16/stateful-false-4        	 1000000	      1016 ns/op
BenchmarkEvents/subs-4/emits-64/stateful-false-4        	 2000000	       965 ns/op
BenchmarkEvents/subs-4/emits-256/stateful-false-4       	 2000000	       934 ns/op
BenchmarkEvents/subs-16/emits-1/stateful-false-4        	  500000	      3952 ns/op
BenchmarkEvents/subs-16/emits-4/stateful-false-4        	  500000	      2934 ns/op
BenchmarkEvents/subs-16/emits-16/stateful-false-4       	  500000	      2725 ns/op
BenchmarkEvents/subs-16/emits-64/stateful-false-4       	  500000	      2758 ns/op
BenchmarkEvents/subs-16/emits-256/stateful-false-4      	  500000	      2617 ns/op
BenchmarkEvents/subs-64/emits-1/stateful-false-4        	  100000	     14211 ns/op
BenchmarkEvents/subs-64/emits-4/stateful-false-4        	  200000	      9108 ns/op
BenchmarkEvents/subs-64/emits-16/stateful-false-4       	  200000	      8393 ns/op
BenchmarkEvents/subs-64/emits-64/stateful-false-4       	  200000	      8749 ns/op
BenchmarkEvents/subs-64/emits-256/stateful-false-4      	  200000	      8824 ns/op
BenchmarkEvents/subs-256/emits-1/stateful-false-4       	   30000	     56793 ns/op
BenchmarkEvents/subs-256/emits-4/stateful-false-4       	   50000	     33185 ns/op
BenchmarkEvents/subs-256/emits-16/stateful-false-4      	   50000	     31275 ns/op
BenchmarkEvents/subs-256/emits-64/stateful-false-4      	   50000	     30681 ns/op
BenchmarkEvents/subs-256/emits-256/stateful-false-4     	   50000	     30056 ns/op
BenchmarkEvents/subs-1/emits-1/stateful-true-4          	 3000000	       445 ns/op
BenchmarkEvents/subs-1/emits-4/stateful-true-4          	 3000000	       460 ns/op
BenchmarkEvents/subs-1/emits-16/stateful-true-4         	 3000000	       498 ns/op
BenchmarkEvents/subs-1/emits-64/stateful-true-4         	 3000000	       458 ns/op
BenchmarkEvents/subs-1/emits-256/stateful-true-4        	 3000000	       457 ns/op
BenchmarkEvents/subs-4/emits-1/stateful-true-4          	 1000000	      1531 ns/op
BenchmarkEvents/subs-4/emits-4/stateful-true-4          	 1000000	      1416 ns/op
BenchmarkEvents/subs-4/emits-16/stateful-true-4         	 1000000	      1022 ns/op
BenchmarkEvents/subs-4/emits-64/stateful-true-4         	 2000000	       968 ns/op
BenchmarkEvents/subs-4/emits-256/stateful-true-4        	 2000000	       918 ns/op
BenchmarkEvents/subs-16/emits-1/stateful-true-4         	  500000	      4004 ns/op
BenchmarkEvents/subs-16/emits-4/stateful-true-4         	  500000	      2946 ns/op
BenchmarkEvents/subs-16/emits-16/stateful-true-4        	  500000	      2762 ns/op
BenchmarkEvents/subs-16/emits-64/stateful-true-4        	  500000	      2725 ns/op
BenchmarkEvents/subs-16/emits-256/stateful-true-4       	  500000	      2595 ns/op
BenchmarkEvents/subs-64/emits-1/stateful-true-4         	  100000	     14315 ns/op
BenchmarkEvents/subs-64/emits-4/stateful-true-4         	  200000	      9137 ns/op
BenchmarkEvents/subs-64/emits-16/stateful-true-4        	  200000	      8480 ns/op
BenchmarkEvents/subs-64/emits-64/stateful-true-4        	  200000	      8511 ns/op
BenchmarkEvents/subs-64/emits-256/stateful-true-4       	  200000	      8800 ns/op
BenchmarkEvents/subs-256/emits-1/stateful-true-4        	   30000	     57276 ns/op
BenchmarkEvents/subs-256/emits-4/stateful-true-4        	   50000	     33244 ns/op
BenchmarkEvents/subs-256/emits-16/stateful-true-4       	   50000	     30441 ns/op
BenchmarkEvents/subs-256/emits-64/stateful-true-4       	   50000	     29496 ns/op
BenchmarkEvents/subs-256/emits-256/stateful-true-4      	   50000	     29768 ns/op
```

**Comment by Kubuxu** [comment]
As a note, this run is not representative as it was run on my notebook in power save.

File: basic_test.go [hunk 1]
```diff
@@ -363,69 +363,89 @@ func TestBothMany(t *testing.T) {
...
-
-func BenchmarkMs1e2m4(b *testing.B) {
-	b.N = 1000000
-	b.ReportAllocs()
-	testMany(b, 10, 100, 10000, false)
+func genTestCases() []benchCase {
+	ret := make([]benchCase, 0, 200)
+	for stateful := 0; stateful < 2; stateful++ {
+		for subs := uint(0); subs <= 8; subs = subs + 2 {
+			for emits := uint(0); emits <= 8; emits = emits + 2 {
```
Comment (magik6k) [inline basic_test.go]: I'd do 
```suggestion
			for emits := uint(0); emits <= 8; emits = emits + 4 {
```
On both of these.

**Comment by Kubuxu** [comment]
```
BenchmarkEvents/subs-001/emits-001/stateful-false-4         	 3000000	       419 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-001/emits-016/stateful-false-4         	 3000000	       468 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-001/emits-256/stateful-false-4         	 3000000	       459 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-016/emits-001/stateful-false-4         	  500000	      4060 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-016/emits-016/stateful-false-4         	  500000	      2742 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-016/emits-256/stateful-false-4         	  500000	      2630 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-256/emits-001/stateful-false-4         	   30000	     57297 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-256/emits-016/stateful-false-4         	   50000	     31604 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-256/emits-256/stateful-false-4         	   50000	     32519 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-001/emits-001/stateful-true-4          	 3000000	       401 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-001/emits-016/stateful-true-4          	 3000000	       484 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-001/emits-256/stateful-true-4          	 3000000	       467 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-016/emits-001/stateful-true-4          	  500000	      4145 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-016/emits-016/stateful-true-4          	  500000	      2937 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-016/emits-256/stateful-true-4          	  500000	      3138 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-256/emits-001/stateful-true-4          	   20000	     60555 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-256/emits-016/stateful-true-4          	   30000	     37719 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-256/emits-256/stateful-true-4          	   50000	     31895 ns/op	       0 B/op	       0 allocs/op
```

**Comment by magik6k** [comment]
Also, allocation reporting doesn't appear to work (but timing matches master tests):
```
goos: linux
goarch: amd64
pkg: github.com/libp2p/go-eventbus
BenchmarkEvents/subs-1/emits-1/stateful-false-32         	10000000	       188 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-1/emits-16/stateful-false-32        	 3000000	       590 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-1/emits-256/stateful-false-32       	 1000000	      1072 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-16/emits-1/stateful-false-32        	  200000	      8979 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-16/emits-16/stateful-false-32       	  500000	      3110 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-16/emits-256/stateful-false-32      	  500000	      2440 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-256/emits-1/stateful-false-32       	   10000	    157449 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-256/emits-16/stateful-false-32      	  100000	     22495 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-256/emits-256/stateful-false-32     	  100000	     14126 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-1/emits-1/stateful-true-32          	10000000	       201 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-1/emits-16/stateful-true-32         	 2000000	       614 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-1/emits-256/stateful-true-32        	 1000000	      1111 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-16/emits-1/stateful-true-32         	  200000	      8377 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-16/emits-16/stateful-true-32        	  500000	      3013 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-16/emits-256/stateful-true-32       	  500000	      2589 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-256/emits-1/stateful-true-32        	   10000	    155322 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-256/emits-16/stateful-true-32       	  100000	     22872 ns/op	       0 B/op	       0 allocs/op
BenchmarkEvents/subs-256/emits-256/stateful-true-32      	  100000	     13952 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/libp2p/go-eventbus	34.801s

```

**Comment by Kubuxu** [comment]
Those benchmarks on purpose have setup removed (using ResetTimer).

**Pull request merged**